### PR TITLE
feat: Add Repair Policy cloud provider interface

### PR DESCRIPTION
**Pull request by engedaam**
Fixes #N/A 

**Description**
- Implementation repair after policy with Kubelet node conditions 
- Current supported condition:
    - Ready
    - DiskPresure 
    - MemoryPressure

**How was this change tested?**

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

**Review by engedaam — COMMENTED** [review]
/karpetner snapshot

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: test/suites/integration/repair_policy_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		env.ExpectCreated(nodeClass, nodePool, dep)
+		pod := env.EventuallyExpectHealthyPodCount(selector, numPods)[0]
+		node := env.ExpectCreatedNodeCount("==", 1)[0]
+		env.EventuallyExpectInitializedNodeCount("==", 1)
+
+		common.ReplaceNodeConditions(node, unhealthyCondition)
+		env.ExpectStatusUpdated(node)
+
+		env.EventuallyExpectNotFound(pod, node)
+		env.EventuallyExpectHealthyPodCount(selector, numPods)
```
Comment (rschalo) [inline test/suites/integration/repair_policy_test.go:68]: Are we also expecting a number of healthy pods?

File: pkg/cloudprovider/cloudprovider.go [hunk 2]
```diff
@@ -246,6 +246,27 @@ func getTags(ctx context.Context, nodeClass *v1.EC2NodeClass, nodeClaim *karpv1.
 	}), staticTags)
 }
 
+func (c *CloudProvider) RepairPolicies() []cloudprovider.RepairPolicy {
```
Comment (rschalo) [inline pkg/cloudprovider/cloudprovider.go:249]: If these are all common to core kubernetes, do these belong in `kubernetes-sigs/karpenter` and then those policies could be extended?

File: test/pkg/environment/common/expectations.go [hunk 3]
```diff
@@ -89,6 +89,36 @@ func (env *Environment) ExpectUpdated(objects ...client.Object) {
...
+func (env *Environment) ExpectStatusUpdated(objects ...client.Object) {
+	GinkgoHelper()
+	for _, o := range objects {
+		Eventually(func(g Gomega) {
+			current := o.DeepCopyObject().(client.Object)
+			g.Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(current), current)).To(Succeed())
+			if current.GetResourceVersion() != o.GetResourceVersion() {
+				log.FromContext(env).Info(fmt.Sprintf("detected an update to an object (%s) with an outdated resource version, did you get the latest version of the object before patching?", lo.Must(apiutil.GVKForObject(o, env.Client.Scheme()))))
+			}
+			o.SetResourceVersion(current.GetResourceVersion())
```
Comment (rschalo) [inline test/pkg/environment/common/expectations.go:107]: Why log here?

File: test/suites/integration/repair_policy_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			{
+				Nodes: "0",
+			},
+		}
+		env.ExpectCreated(nodeClass, nodePool, dep)
+		pod := env.EventuallyExpectHealthyPodCount(selector, numPods)[0]
+		node := env.ExpectCreatedNodeCount("==", 1)[0]
+		env.EventuallyExpectInitializedNodeCount("==", 1)
+
+		unhealthyCondition := corev1.NodeCondition{
```
Comment (rschalo) [inline test/suites/integration/repair_policy_test.go:97]: Can we add this at the top and then use it in each of these tests?

File: test/pkg/environment/common/expectations.go (hunk 3 above)
Comment (jonathan-innis) [reply test/pkg/environment/common/expectations.go:107]: +1

File: test/suites/integration/repair_policy_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		selector = labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)
+	})
+
+	DescribeTable("Conditions", func(unhealthyCondition corev1.NodeCondition) {
+		env.ExpectCreated(nodeClass, nodePool, dep)
+		pod := env.EventuallyExpectHealthyPodCount(selector, numPods)[0]
+		node := env.ExpectCreatedNodeCount("==", 1)[0]
+		env.EventuallyExpectInitializedNodeCount("==", 1)
+
+		common.ReplaceNodeConditions(node, unhealthyCondition)
```
Comment (jonathan-innis) [inline test/suites/integration/repair_policy_test.go:64]:
```suggestion
		node = common.ReplaceNodeConditions(node, unhealthyCondition)
```

nit: From what I can tell, this does update the status conditions in-place, but it seems a little easier to reason about if we use the return object

**Review by jonathan-innis — DISMISSED** [review]

File: pkg/cloudprovider/cloudprovider.go (hunk 2 above)
Comment (engedaam) [reply pkg/cloudprovider/cloudprovider.go:249]: We don't want to enforce condition on cloud providers or prescribe the tolerationDuration amount for a cloud provider for this alpha feature.

File: test/suites/integration/repair_policy_test.go (hunk 1 above)
Comment (engedaam) [reply test/suites/integration/repair_policy_test.go:68]: No these tests are only expect there to be one healthy node that goes bad

File: test/pkg/environment/common/expectations.go (hunk 3 above)
Comment (engedaam) [reply test/pkg/environment/common/expectations.go:107]: This is how we have been doing our update call traditionally, so I was keep with that custom:

```
func (env *Environment) ExpectUpdated(objects ...client.Object) {
	GinkgoHelper()
	for _, o := range objects {
		Eventually(func(g Gomega) {
			current := o.DeepCopyObject().(client.Object)
			g.Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(current), current)).To(Succeed())
			if current.GetResourceVersion() != o.GetResourceVersion() {
				log.FromContext(env).Info(fmt.Sprintf("detected an update to an object (%s) with an outdated resource version, did you get the latest version of the object before patching?", lo.Must(apiutil.GVKForObject(o, env.Client.Scheme()))))
			}
			o.SetResourceVersion(current.GetResourceVersion())
			g.Expect(env.Client.Update(env.Context, o)).To(Succeed())
		}).WithTimeout(time.Second * 10).Should(Succeed())
	}
}
```
This help track were we see any conflicts. If you think this is not need I can drop it

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Pull request merged**